### PR TITLE
Feature - 288 cron task to delete fv records

### PR DESCRIPTION
**Pull request by omerzam**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

- Added a new Mongoose schema for new face_verification table
- Added new FV mongoose model to user private provider
- upsert FV record with new date onFaceVerificationComplete
- new cron task to remove FV records older than authenticationPeriod set in identity contract

About # [288](https://app.zenhub.com/workspaces/gooddollar-5d50833888a83c6880d3e345/issues/gooddollar/goodserver/288)

# How Has This Been Tested?

Unitests

# Checklist:
- [X] PR title matches follow: (Feature|Bug|Chore) Task Name
- [X] My code follows the style guidelines of this project
- [X] I have followed all the instructions described in the initial task (check Definitions of Done)
- [X] I have performed a self-review of my own code
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] I have added reference to a related issue in the repository
- [X] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request (for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: src/server/db/__tests__/UserPrivate.js [hunk 1]
```diff
@@ -193,6 +194,27 @@ describe('UserPrivate', () => {
...
+    //   lastFVDate: now,
+    //   enrollmentIdentifier
+    // }
+
+    await storage.upsertFaceVerificationRecord(enrollmentIdentifier)
+    const result = await faceVerificationsModel.findOne({ enrollmentIdentifier })
+    expect(result.enrollmentIdentifier).toEqual(enrollmentIdentifier)
+  })
+
+  it('Should get a Face Verifications record ny enrollmentIdentifier', async () => {
```
Comment (sirpy) [inline src/server/db/__tests__/UserPrivate.js:210]:
```suggestion
  it('Should get a Face Verifications record by enrollmentIdentifier', async () => {
```

File: src/server/db/mongo/models/faceVerifications.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import mongoose, { Schema } from '../../mongo-db.js'
+import { MODEL_FACE_VERIFICATIONS } from './constants'
+
+export const FaceVerificationsSchema = new Schema({
+  lastFVDate: {
+    type: Date
```
Comment (sirpy) [inline src/server/db/mongo/models/faceVerifications.js:6]:
```suggestion
    type: Date
    index: true
```

File: src/server/db/mongo/models/faceVerifications.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  },
+  enrollmentIdentifier: {
+    type: String
```
Comment (sirpy) [inline src/server/db/mongo/models/faceVerifications.js:9]:
```suggestion
    type: String,
    index: { unique: true }
```

File: src/server/db/mongo/models/faceVerifications.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  }
+})
+FaceVerificationsSchema.index({ lastFVDate: 1 }, { unique: true }) // schema level
+FaceVerificationsSchema.index({ enrollmentIdentifier: 1 }, { unique: true }) // schema level
```
Comment (sirpy) [inline src/server/db/mongo/models/faceVerifications.js:13]:
```suggestion
```

File: src/server/verification/verificationAPI.js [hunk 5]
```diff
@@ -139,7 +139,7 @@ const setup = (app: Router, verifier: VerificationAPI, gunPublic: StorageAPI, st
         // if user is already verified, we're skipping enroillment logic
         if (user.isVerified || disableFaceVerification || allowDuplicatedFaceRecords || isE2ERunning) {
           // creating enrollment session manually for this user
-          const enrollmentSession = enrollmentProcessor.createEnrollmentSession(user, log)
+          const enrollmentSession = enrollmentProcessor.createEnrollmentSession(user, enrollmentIdentifier, log)
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js:142]: there's probably not reason to add enrollmentIdentifier since it is probably accessible via the next code line
initialize(payload)

**Review by sirpy — CHANGES_REQUESTED** [review]
where is the crontask itself?

File: src/server/db/mongo/models/faceVerifications.js (hunk 2 above)
Comment (omerzam) [reply src/server/db/mongo/models/faceVerifications.js:6]: i dont see this in the documentation

File: src/server/blockchain/AdminWallet.js [hunk 6]
```diff
@@ -368,7 +368,7 @@ export class Wallet {
     }
   }
 
-  async getAuthenticationPeriod(): Promise<TransactionReceipt> {
+  async getAuthenticationPeriod(): Promise<number> {
     try {
       const result = await this.identityContract.methods.authenticationPeriod().call()
       return result.toString()
```
Comment (sirpy) [inline src/server/blockchain/AdminWallet.js:374]:
```suggestion
      return result.toNumber()
```

File: src/server/verification/cron/DisposeEnrollmentsPeriodic.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  async execute({ setTime }) {
+    try {
+      logger.info('Starting worker', this.name)
+      const authenticationPeriod = await this.getAuthenticationPeriod()
+      logger.info('Authentication period found', authenticationPeriod)
+      const fvRecords = await this.getFVRecords(authenticationPeriod)
+      logger.info(`Fetched ${fvRecords.length} FV records from db`)
+      const disposeEnrollments = await this.disposeEnrollments(fvRecords)
+      logger.info(`Performed Dispose Enrollment for ${disposeEnrollments.length} enrollmentIds`)
+      const enrollmentIdentifiers = disposeEnrollments.reduce((arr, disposeEnrollment) => {
```
Comment (sirpy) [inline src/server/verification/cron/DisposeEnrollmentsPeriodic.js:30]: use .filter

File: src/server/verification/cron/DisposeEnrollmentsPeriodic.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+   */
+  async disposeEnrollments(fvRecords) {
+    const enrollmentIds = fvRecords.map(fvRecord => {
+      return fvRecord.enrollmentIdentifier
+    })
+
+    await AdminWallet.ready
+    logger.info('Enrollment Ids to dispose:', enrollmentIds)
+    const enrollmentProcessor = createEnrollmentProcessor(UserDBPrivate, logger)
+    const disposedEnrollments = await enrollmentProcessor.disposeEnrollments(enrollmentIds)
```
Comment (sirpy) [inline src/server/verification/cron/DisposeEnrollmentsPeriodic.js:67]: does this immediatly delete the records?

File: src/server/verification/cron/DisposeEnrollmentsPeriodic.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+      logger.info('Authentication period found', authenticationPeriod)
+      const fvRecords = await this.getFVRecords(authenticationPeriod)
+      logger.info(`Fetched ${fvRecords.length} FV records from db`)
+      const disposeEnrollments = await this.disposeEnrollments(fvRecords)
+      logger.info(`Performed Dispose Enrollment for ${disposeEnrollments.length} enrollmentIds`)
+      const enrollmentIdentifiers = disposeEnrollments.reduce((arr, disposeEnrollment) => {
+        if (disposeEnrollment.status !== 'failed') arr.push(disposeEnrollment.enrollmentIdentifier)
+        return arr
+      }, [])
+      logger.info('Removing disposed FV Rcords from DB')
```
Comment (sirpy) [inline src/server/verification/cron/DisposeEnrollmentsPeriodic.js:34]: log how many records are being delete
if records failed dispose log.warn number of records failed + their ids

File: src/server/verification/cron/DisposeEnrollmentsPeriodic.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+  /**
+   * restore trust profiles
+   */
+  async disposeEnrollments(fvRecords) {
+    const enrollmentIds = fvRecords.map(fvRecord => {
+      return fvRecord.enrollmentIdentifier
+    })
+
+    await AdminWallet.ready
```
Comment (sirpy) [inline src/server/verification/cron/DisposeEnrollmentsPeriodic.js:64]: why is this here?

**Review by sirpy — CHANGES_REQUESTED** [review]

File: src/server/verification/verificationAPI.js (hunk 5 above)
Comment (omerzam) [reply src/server/verification/verificationAPI.js:142]: it is not in the payload it is in the params

File: src/server/verification/cron/DisposeEnrollmentsPeriodic.js (hunk 7 above)
Comment (omerzam) [reply src/server/verification/cron/DisposeEnrollmentsPeriodic.js:30]: using filter will produce an array of objects with enrollmentIdentifier and status, what i wanted is an array of enrollmentIdentifiers to send as an argument to removeRecords method

File: src/server/verification/cron/DisposeEnrollmentsPeriodic.js (hunk 8 above)
Comment (omerzam) [reply src/server/verification/cron/DisposeEnrollmentsPeriodic.js:67]: no, this is just disposing them from the provider (zoom) and returning the status of each identifier (sucess, failed, notExists
deleting the records from DB is done after at the removeRecords method and not to the failed ones

File: src/server/verification/cron/DisposeEnrollmentsPeriodic.js (hunk 9 above)
Comment (omerzam) [reply src/server/verification/cron/DisposeEnrollmentsPeriodic.js:34]: there is logging for this information at the disposeEnrollments method at EnrollmentProcessor class
if more logging is needed I will add

Comment (sirpy) [reply src/server/verification/cron/DisposeEnrollmentsPeriodic.js]: add the number of records failed

**Review by serdiukov-o-nordwhale — CHANGES_REQUESTED** [review]
all that stuff isn't needed because actually we already have the tasks queue to dispose enrollments.

How i see the solution:
1. remove `createdAt` from `DelayedTaskSchema`. we will use instead `scheduledAt` Date field with ts when it actually should be executed. We could set any value (now to exec immediately, now + 24h, now + 2 weeks etc)
2. subject field by design could have any value depending of task type. for dispose enrollments it's just a string with enrollmend id now. Need to change it to the object 
```javascript
{ enrollmentIdentifier: string, scheduledAt: Date, isDisposing?: boolean }
```
3. add `cancelTasksQueued` method to the UserPrivateDB with the interface similar to `hasTasksQueued` (but returning `Promise<void>`). It should remove tasks of taskName and filters specified
4. add `taskUtil.js` file, move `DISPOSE_ENROLLMENTS_TASK` constant definition and export function `cancelEnrollmentDisposal(storage, enrollmentIdentifier)` which calls
```javascript
cancelTasksQueued(DISPOSE_ENROLLMENTS_TASK, { 'subject.enrollmentIdentifier': enrollmentIdentifier })
```
5. in `EnrollmentSession` / `onEnrollmentCompleted` add 
```javascript
const authenticationPeriod = await adminApi.getAuthenticationPeriod(); 
const disposeAt = moment().add(authenticationPeriod, 'days').toDate();

await storage.cancelEnrollmentDisposal(enrollmentIdentifier);  // remove existing tasks if any
await storage.enqueueTask(DISPOSE_ENROLLMENTS_TASK, { enrollmentIdentifier, disposeAt })
```
this will add a task to auto remove FV record after authenticationPeriod days passed
6. in `enqueueDisposal` update `enqueueTask`  call to

```javascript
let disposeAt = moment(); 

if (keepEnrollments > 0) { 
  disposeAt = disposeAt.add(keepEnrollments, 'hours'); 
} 

disposeAt = disposeAt.toDate(); 
await storage.cancelEnrollmentDisposal(enrollmentIdentifier); 
await storage.enqueueTask(DISPOSE_ENROLLMENTS_TASK, { enrollmentIdentifier, disposeAt, isDisposing: true })
```

this will add task to be executed ASAP if `keepEnrollments` set to 0 or will add 24h removal. `isDisposing` flag is needed to differ it from authenticationPeriod removal
7. update hasTasksQueued call in isEnqueuedForDisposal to 
```javascript
hasTasksQueued(DISPOSE_ENROLLMENTS_TASK, { subject: { enrollmentIdentifier, isDisposing: true })
```
8. update `disposeEnqueuedEnrollments`. Remove `if (keepEnrollments > 0)...`  condition block. rename enqueuedAtFilters to disposeAtFilters and define it as 
```javascript
const disposeAtFilters = { 'subject.disposeAt': { $lte: moment().toDate() }} 
```
(or just new Date())
this will select all tasks (both on account removal and on authenticationPeriod has passed)
